metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractValueCodeAction.scala

```diff
@@ -0,0 +1,311 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.annotation.tailrec
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Defn
+import scala.meta.Enumerator
+import scala.meta.Name
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.internal.trees.Origin.Parsed
+import scala.meta.pc.CancelToken
+import scala.meta.tokens.Token
+
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.{lsp4j => l}
+
+class ExtractValueCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorExtract
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val path = params.getTextDocument().getUri().toAbsolutePath
+    val range = params.getRange()
+
+    val applyOpt = trees.findLastEnclosingAt[Term.Apply](
```

Comment:
I'm not sure if it's right here or not that ends up being the issue, but I've found that at times it ends up not extracting what I'd expect. For example I tried this on various places in the Metals code base.

```scala
  private def newSymbolIndex(): OnDemandSymbolIndex = {
    OnDemandSymbolIndex.empty(
      onError = {
        case e @ (_: ParseException | _: TokenizeException) =>
          scribe.error(e.toString)
        case e: InvalidJarException =>
          scribe.warn(s"invalid jar: ${e.path}")
        case _: NoSuchFileException =>
        // only comes for badly configured jar with `/Users` path added.
        case NonFatal(e) =>
          scribe.error("unexpected error during source scanning", e)
      },
      toIndexSource = path => {
        if (path.is<<>>AmmoniteScript)
          for {
            target <- buildTargets.sourceBuildTargets(path).headOption
            toIndex <- ammonite.generatedScalaPath(target, path)
          } yield toIndex
        else
          None
      }
    )
  }
```

I sort of expected this to extract `path.isAmmoniteScript`, but instead it extracted:

```scala
    val newValue = toIndexSource = path => {
        if (path.isAmmoniteScript)
          for {
            target <- buildTargets.sourceBuildTargets(path).headOption
            toIndex <- ammonite.generatedScalaPath(target, path)
          } yield toIndex
        else
          None
      }
```

Reply:
I forgot about named arguments! :fearful: 

The current approach will try to extract a parameter, it doesn't work for if conditions or anything else. We could do that for sure, but maybe we could refactor it later?

Reply:
For sure!

Reply:
I fixed the issue with named parameters and added a test for it (modified existing one to use named params).

I have an idea of how to add support for Term.If etc, but what do you think we would want to extract also?

I think this can be:
- condition in while/if/doWhile
- interpolate
- return
- throw
- tuple
- matched expression
- new

Anything else comes to mind ?

Reply:
🤔 I'm not really sure just off the top of my head. This is probably one of those things that I'd want to use and play around with, try to extract something, not be able to and then know that I want it 😆 I also think it's fine to start small and add as we go.